cudamapper/src/chainer_utils.cuh

```diff
@@ -0,0 +1,149 @@
+///
+/// \param start The first anchor in the chain.
+/// \param end The last anchor in the chain.
+/// \param num_anchors The total number of anchors in the chain.
+__host__ __device__ Overlap create_simple_overlap(const Anchor& start,
+                                                  const Anchor& end,
+                                                  const int32_t num_anchors);
+
+///
+/// \brief Perform a backtrace on the predecessors array
+/// to generate overlaps from anchors. Anchors must have been chained
+/// by a chaining function that filles the predecessors and scores array.
+///
+/// \param anchors An array of anchors.
+/// \param overlaps An array of overlaps to be filled.
+/// \param scores An array of scores. Only chains with a score greater than
+/// min_score will be backtraced.
+/// \param max_select_mask A boolean mask, used to mask out any subchains during
+/// the backtrace.
+/// \param predecessors An array of indices into the anchors array marking the predecessor
+/// of each anchor within a chain.
+/// \param n_anchors The number of anchors.
+/// \param min_score The minimum score of a chain for performing backtracing.
+/// \return __global__
+__global__ void backtrace_anchors_to_overlaps(const Anchor* anchors,
+                                              Overlap* overlaps,
+                                              double* scores,
+                                              bool* max_select_mask,
+                                              int32_t* predecessors,
+                                              const int32_t n_anchors,
+                                              const int32_t min_score);
+
+///
+/// \brief Allocate a 1-dimensional array representing an unrolled 2D-array
+/// (overlap X n_anchors_in_overlap) of anchors within each overlap. Rather than
+/// copy the anchors, the final array holds the indices within the anchors array
+/// of the anchors in the chain.
+///
+/// \param overlaps An array of Overlaps. Must have a well-formed num_residues_ field
+/// \param unrolled_anchor_chains  An array of int32_t. Will be resided on return.
+/// \param anchor_chain_starts An array holding the index in the anchors array of the first
+/// anchor in an overlap.
+/// \param num_overlaps The number of overlaps in the overlaps array.
+/// \param num_total_anchors The number of anchors in the anchors array.
+/// \param _allocator The DefaultDeviceAllocator for this overlapper.
+/// \param _cuda_stream The cudastream to allocate memory within.
+void allocate_anchor_chains(device_buffer<Overlap>& overlaps,
+                            device_buffer<int32_t>& unrolled_anchor_chains,
+                            device_buffer<int32_t>& anchor_chain_starts,
+                            int32_t num_overlaps,
+                            int32_t& num_total_anchors,
+                            DefaultDeviceAllocator& _allocator,
+                            cudaStream_t& _cuda_stream);
```

Comment:
Both allocator and stream should be passed by value (allocator is just a wrapper around a `shared_ptr` which points to actual memory resource)

Reply:
done